# Data loading and exploration
Load California Housing dataset

In [1]:
from sklearn.datasets import fetch_california_housing
import pandas as pd

housing = fetch_california_housing()

Create dataframe for features and target variable

In [2]:
X = pd.DataFrame(housing.data, columns=housing.feature_names) 
print(X)
y = pd.Series(housing.target, name='med_house_value')

       MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0      8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1      8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2      7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3      5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4      3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   
...       ...       ...       ...        ...         ...       ...       ...   
20635  1.5603      25.0  5.045455   1.133333       845.0  2.560606     39.48   
20636  2.5568      18.0  6.114035   1.315789       356.0  3.122807     39.49   
20637  1.7000      17.0  5.205543   1.120092      1007.0  2.325635     39.43   
20638  1.8672      18.0  5.329513   1.171920       741.0  2.123209     39.43   
20639  2.3886      16.0  5.254717   1.162264      1387.0  2.616981     39.37   

       Longitude  
0        -122.23  
1

## Initial exploration of the dataset: 
First, display the first 5 rows

In [3]:
# display the first 5 rows
print("Dataset preview: ")
print(X.head())

Dataset preview: 
   MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  
0    -122.23  
1    -122.22  
2    -122.24  
3    -122.25  
4    -122.25  


Display the names of features and check for any NA values

In [4]:
# give names of features; check for NA
print("\n Feature names: ")
print(list(X.columns))
print("\n Check for NA values by variable: ")
print(X.isna().sum())


 Feature names: 
['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude']

 Check for NA values by variable: 
MedInc        0
HouseAge      0
AveRooms      0
AveBedrms     0
Population    0
AveOccup      0
Latitude      0
Longitude     0
dtype: int64


Show some basic summary statistics:

In [5]:
# summary stats
print("\n Basic summary statistics: ")
print(X.describe())


 Basic summary statistics: 
             MedInc      HouseAge      AveRooms     AveBedrms    Population  \
count  20640.000000  20640.000000  20640.000000  20640.000000  20640.000000   
mean       3.870671     28.639486      5.429000      1.096675   1425.476744   
std        1.899822     12.585558      2.474173      0.473911   1132.462122   
min        0.499900      1.000000      0.846154      0.333333      3.000000   
25%        2.563400     18.000000      4.440716      1.006079    787.000000   
50%        3.534800     29.000000      5.229129      1.048780   1166.000000   
75%        4.743250     37.000000      6.052381      1.099526   1725.000000   
max       15.000100     52.000000    141.909091     34.066667  35682.000000   

           AveOccup      Latitude     Longitude  
count  20640.000000  20640.000000  20640.000000  
mean       3.070655     35.631861   -119.569704  
std       10.386050      2.135952      2.003532  
min        0.692308     32.540000   -124.350000  
25%      

# Linear Regression on Unscaled Data

Set up the linear regression with testing and training data and make first predictions

In [6]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

# split 80-20 train test
X_train_raw, X_test_raw, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=20)
# can use random_state=int to control seed

# set up linear regression
lin_reg_raw = LinearRegression()
lin_reg_raw.fit(X_train_raw, y_train)

# predict from training
y_pred_raw = lin_reg_raw.predict(X_test_raw)

Evaluate the model performance with $MSE$, $RMSE$, and $R^2$

In [7]:
from sklearn.metrics import mean_squared_error, root_mean_squared_error, r2_score

mse_raw = mean_squared_error(y_test, y_pred_raw)
rmse_raw = root_mean_squared_error(y_test, y_pred_raw)
r2_raw = r2_score(y_test, y_pred_raw)

print("Metrics for model performance: ")
print(f"MSE: {mse_raw:.4f}")
print(f"RMSE: {rmse_raw:.4f}")
print(f"R² Score: {r2_raw:.4f}")

print("\n Model coefficients: ")
print(pd.Series(lin_reg_raw.coef_,
                index=X.columns))

Metrics for model performance: 
MSE: 0.5410
RMSE: 0.7355
R² Score: 0.6122

 Model coefficients: 
MedInc        0.438372
HouseAge      0.009309
AveRooms     -0.117517
AveBedrms     0.648578
Population   -0.000007
AveOccup     -0.003417
Latitude     -0.413687
Longitude    -0.426866
dtype: float64


### What does the R² score tell us about model performance?

The calculated R² score is 0.6122, which indicates that the model is able to explain over half of the variance - a reasonably good fit.

### Which features seem to have the strongest impact on predictions based on the model’s coefficients?

Going solely off of the value of estimated coefficients, it seems as though the average number of bedrooms and median income have the largest effect on the house value. Intuitively, this makes sense, as median income likely reflects strong neighborhood amenities, and the number of bedrooms is increasing in house size.

### How well do the predicted values match the actual values?

The root mean squared error for the model is 0.7355, which gives an idea of the degree to which the predicted values deviate from the actual values in units of y. Given that the mean y value is approximately 2 with standard deviation of 1.15, the predicted values do not seem to match the actual values extremely well.


# Feature Selection and Simplified Model

I select median income, average rooms, and average bedrooms to be part of the simplified model, as these have the largest coefficients by magnitude. Without knowing anything else about the coefficients (e.g. SE or t-stat) it is impossible to know how precise these coefficients are, so I use a simple selection based on size.

In [10]:
# create subset of X data and repeat the same analysis as above
X_sub = X[['MedInc','AveRooms','AveBedrms']]

X_train_sub, X_test_sub, y_train_sub, y_test_sub = train_test_split(X_sub, y, test_size = 0.2, random_state = 20)

lin_reg_sub = LinearRegression()
lin_reg_sub.fit(X_train_sub, y_train_sub)
y_pred_sub = lin_reg_sub.predict(X_test_sub)

mse_sub = mean_squared_error(y_test_sub, y_pred_sub)
rmse_sub = root_mean_squared_error(y_test_sub, y_pred_sub)
r2_sub = r2_score(y_test_sub, y_pred_sub)

print("Metrics for simplified model performance: ")
print(f"MSE: {mse_sub:.4f}")
print(f"RMSE: {rmse_sub:.4f}")
print(f"R² Score: {r2_sub:.4f}")

print("\n Model coefficients: ")
print(pd.Series(lin_reg_sub.coef_,
                index=X_sub.columns))

Metrics for simplified model performance: 
MSE: 0.7010
RMSE: 0.8372
R² Score: 0.4975

 Model coefficients: 
MedInc       0.542090
AveRooms    -0.260379
AveBedrms    1.101575
dtype: float64


One can directly compare the 3 calculated performance metrics between models: the simplified model has both a higher MSE and RMSE, as well as a lower R². This the fitted values are, on average, further away from the real values of y (house value), and the model is able to capture less of the variance in y. In practice, it may be easier to use this model, as collecting data for 3 variables is likely far easier than the full set of 8: this may be particularly relevant if the data source only contains some subset of variables (e.g. a realty data containing data on rooms but not the population of the area).